### PR TITLE
Add API to the GC runtime

### PR DESCRIPTION
**Pull request by jacob-hughes**
I've split this up to be reviewed separately because I think our entry points into the GC from the standard library, user code, and our modified LLVM IR, are going to be fundamental in how the rest of the collector looks. 

I have created a collector as a static, mutable singleton struct which contains a bunch of bookkeeping information about the heap and rootset, and will contain all the methods needed for use within the GC runtime. 

In this draft PR, I'd like to focus solely on the interaction with the runtime externally, so I've not included `collector.rs` as I don't want its implementation details to leak into how we think about the higher level API. However, if you think it might be useful and help give you a bigger picture, let me know, and I'll push it.

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+use collector::COLLECTOR;
+
+/// This must be called before the GC can be used (usually in the setup code
+/// before `main()`). Initialisation consists of two stages:
+///     1. Read the stackmap section in the ELF file into an in-memory table for
+///        fast lookup.
+///     2. Allocate a chunk of heap memory to be used to store objects managed
+///        by the GC.
+pub fn init() {
+        COLLECTOR.alloc_heap().make_root_table();
```
Comment (ltratt) [inline src/api.rs]: I think this file hasn't gone through `rustfmt` ;)

File: src/api.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+//! This file contains the API for the runtime GC. Interaction with the GC is
+//! only legal through these functions. 
+//!
+//! The GC runtime consists of a mutable singleton `Collector` struct:
```
Comment (ltratt) [inline src/api.rs]: I'm not saying we shouldn't do this, but I couldn't immediately work out how we might at some point generalise this to multiple threads. I wonder if the `Collector` struct should appear immutable but have `UnsafeCell`s inside?

File: src/api.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+/// 
+/// No information about whether a poll resulted in a collection is returned to
+/// the mutator. The only thing that can be guaranteed is that a collection
+/// *might* have happened after returning from this call.
+///
+/// ------------------------------ WARNING -------------------------------------
+/// | We need to be really careful about code which panics inside the collector|
+/// | here. The safepoint poll will *not* be called by native Rust code which  |
+/// | means that panic handling is UB. We should probably abort on panic here. |
+/// ----------------------------------------------------------------------------
```
Comment (vext01) [inline src/api.rs]: Should we use the `abort` panic strategy then?

File: src/api.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#[no_mangle]
+extern pub fn safepoint_poll() {
+    if COLLECTOR.should_collect() {
```
Comment (vext01) [inline src/api.rs]: I wonder why you separated `should_collect` from `collect`?

Why not have one function `collect`, which early returns if there's nothing to do?

File: src/api.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+///
+/// Calling `force_collect` multiple times before a safepoint boundary is
+/// pointless - it won't make a collection happen any faster. It simply sets a
+/// flag in the collector which is checked in order to decide whether to make a
+/// collection.
+pub fn force_collect() {
+    COLLECTOR.collect_next();
+}
+
+
```
Comment (vext01) [inline src/api.rs]: Laurie has already mentioned rustfmt ;)

**Comment by vext01** [comment]
A few superficial comments from me.

I have to say, the quality of the code comments in this PR are really good!

File: src/api.rs (hunk 3 above)
Comment (ltratt) [reply src/api.rs]: This is tricky: as a library, our GC can't really impose panic or unwind on users. Ideally, we have to deal with whatever they select. I agree that unwind feels a bit of a mess in Rust.

File: src/api.rs (hunk 4 above)
Comment (ltratt) [reply src/api.rs]: I think "should_collect" is set at a (potentially) non-safepoint, so you can only do the actual collection when you're in a safepoint.

File: src/api.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/api.rs]: Guilty

File: src/api.rs (hunk 2 above)
Comment (jacob-hughes) [reply src/api.rs]: Perhaps you're right. I did worry about the current approach as it means any method on the collector with a `&mut self` would need to be placed in an unsafe block.

Using an `UnsafeCell` inside a static singleton would mean that the collector would need to be instantiated locally, (inside a `thread_local!` macro). I _think_ this sounds right to me. There will need to be atomic access to the collector, and this will mean that only the collector thread can mutate it.

The Rust code for this isn't very pretty, it would look a bit like this in places:

```rust
thread_local!(static GC: UnsafeCell<usize> = UnsafeCell::new(5));

fn main() {
    // Print the value
    GC.with(|f| {
        println!("{:?}", unsafe {*f.get()})
    });

    // mutate some GC value,
    // Then print the value
    GC.with(|f| {
        unsafe { *f.get() = 10;}
        println!("{:?}", unsafe {*f.get()})
    });
}
```

But at least we'd get finer grain control over where our unsafe boundaries are.

File: src/api.rs (hunk 4 above)
Comment (jacob-hughes) [reply src/api.rs]: Laurie's right, but also, `should_collect()` is a small, trivial flag check method so it can be inlined. Performing the check in `collect()` would potentially mean an additional function call for some particularly performance critical part of the code.

**Comment by jacob-hughes** [comment]
> I have to say, the quality of the code comments in this PR are really good!

@vext01 Thanks! I think it's particularly important for this part of the API because each call assumes a bunch of invariants are upheld with no real way to know about this without writing a bunch of  "this is UB / this is fine" requirements in the comments.

File: src/api.rs (hunk 3 above)
Comment (jacob-hughes) [reply src/api.rs]: I'm not sure I have a good answer to this one. At the very least, when debugging, we'll want abort on panic so that we can have some idea what went wrong. But yes, for the user we would need to hopefully recover in some way.

Comment (vext01) [reply src/api.rs]: Can the user really recover if the GC fails?

Comment (ltratt) [reply src/api.rs]: Yes. You can sometimes free up inessential resources (e.g. caches). It doesn't happen often though. Anyway, that's an orthogonal issue: in the long run we *have* to support unwinding, even if we don't like it very much.

File: src/api.rs (hunk 2 above)
Comment (ltratt) [reply src/api.rs]: What's `with` do?

Comment (vext01) [reply src/api.rs]: It grants safe access to the thread local:
> The with method yields a reference to the contained value which cannot be sent across threads or escape the given closure.

https://doc.rust-lang.org/std/thread/struct.LocalKey.html

File: src/api.rs (hunk 3 above)
Comment (vext01) [reply src/api.rs]: I'm sceptical, but let's see what happens.

File: src/api.rs (hunk 2 above)
Comment (ltratt) [reply src/api.rs]: `GC` has type `UnsafeCell<...>` so I can't see a `with` function. Maybe I'm being an idiot!

Comment (vext01) [reply src/api.rs]: `thread_local!` wraps the value in a `LocalKey`.

> This key uses the fastest possible implementation available to it for the target platform. It is instantiated with the thread_local! macro and the primary method is the with method.

Comment (ltratt) [reply src/api.rs]: Aha, I see! I am an idiot :0

**Comment by jacob-hughes** [comment]
I have now changed the API to use a `thread_local!` to house the main `Collector` struct locally on the GC thread. 

There are also a couple of mechanical commits which simply move `src/api.rs` into `src/lib.rs`.

**Comment by jacob-hughes** [comment]
I messed up the pushing commits here before they were ready so apologies if they're a bit hard to follow. In short, I've moved `src/api.rs` to `src/lib.rs` and also added `collector.rs` where I start to flesh out the basic functionality the collector is expected to have.

This is now ready to review.

**Comment by vext01** [comment]
Fine by me.

File: src/collector.rs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+/// The size of the heap in bytes
+const HSIZE: usize = 1024;
+
+/// The byte alignment of the heap
+const HALIGN: usize = 4;
+
+pub(crate) struct Collector {
+    hptr: UnsafeCell<*mut usize>,
```
Comment (ltratt) [inline src/collector.rs]: Can we not make this `Cell` and use `set` and so on?

Comment (jacob-hughes) [reply src/collector.rs]: Yes, I'm not sure why I left this one as `UnsafeCell` when changing the others to `Cell`. I've pushed a new commit to address this now.

File: src/lib.rs [hunk 7]
```diff
@@ -1,4 +1,99 @@
...
+/// Perform a collection at the next safepoint. Note that this *does not* stop
+/// the world immediately to perform garbage collection. Instead, it guarantees
+/// that once the next safepoint boundary has been reached, the collector will
+/// attempt a collection.
+///
+/// Calling `force_collect` multiple times before a safepoint boundary is
+/// pointless - it won't make a collection happen any faster. It simply sets a
+/// flag in the collector which is checked in order to decide whether to make a
+/// collection.
+pub fn force_collect() {
```
Comment (ltratt) [inline src/lib.rs:77]: I am not sure about `force_collect` -- because it doesn't force a collection AFAICS? Or does it block the current thread until a collection has occurred?

File: src/lib.rs [hunk 8]
```diff
@@ -1,4 +1,99 @@
...
+    COLLECTOR.with(|c| c.collect_next());
+}
+
+/// Stores an object in the GC heap and returns a raw pointer. `alloc_raw`
+/// should not be called directly by the user. Instead, it is exposed so that
+/// the standard library can build a GC smart pointer to a managed object.
+///
+/// Allocation to the GC heap is cheap and consists of: a heap bounds check; the
```
Comment (ltratt) [inline src/lib.rs]: I don't think we should define what an allocation consists of in the public API. For example, in a non-moving collector it probably won't be a pointer bump.

File: src/collector.rs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            hstart: Cell::new(0),
+            hend: Cell::new(0),
+
+            collect_next: Cell::new(false),
+            roots: UnsafeCell::new(None)
+        }
+    }
+
+    #[inline]
+    pub fn collect_next(&self) {
```
Comment (ltratt) [inline src/collector.rs:42]: I'm not sure why, but this name doesn't really have the right sense to me. It's more `please_collect_soon` or similar!

File: src/lib.rs (hunk 7 above)
Comment (jacob-hughes) [reply src/lib.rs:77]: I'm not sure it's the best name either. You're right, that it _doesn't_ force a collection immediately, but it means that the next time the mutator hits a safepoint, the collector will pause to reclaim memory. It won't block the current thread.

File: src/lib.rs (hunk 8 above)
Comment (jacob-hughes) [reply src/lib.rs]: Good point, this is implementation specific and shouldn't be public.

File: src/collector.rs (hunk 9 above)
Comment (jacob-hughes) [reply src/collector.rs:42]: What `collect_next` really means is `this_will_collect_the_next_time_we_hit_a_safepoint_poll`. "Please" implies that there may be a reason that the future collection may be aborted/postponed, but at the moment, that's not the case.

File: src/lib.rs (hunk 7 above)
Comment (ltratt) [reply src/lib.rs:77]: Right, so we need to change the name to something that doesn't suggest that it forces a GC. I'd suggest something like `request_collect_soon` which is a bit of a mouthful, but at least not easily misunderstood.

Comment (jacob-hughes) [reply src/lib.rs:77]: I've changed this to `request_collect_soon`, but I don't really like it. I'm not sure what would be better though.

**Comment by jacob-hughes** [comment]
Ready for re-review. I've also changed `alloc_raw` to return a result type instead of a raw pointer. This way we can leave it up to the library user to decide what to do on issues such as running out of memory.

File: src/lib.rs (hunk 7 above)
Comment (ltratt) [reply src/lib.rs:77]: Maybe `queue_collect` or `schedule_collect` or similar?

File: src/lib.rs (hunk 8 above)
Comment (ltratt) [reply src/lib.rs]: Don't forget this bit.

**Comment by jacob-hughes** [comment]
[As discussed in person] I've changed the API so that `force_collect` actually blocks to collect when called.

**Comment by jacob-hughes** [comment]
Ready for re-review

**Comment by ltratt** [comment]
Ready to squash?

**Comment by jacob-hughes** [comment]
Squashed
